### PR TITLE
fdc3 2.0 workbench update

### LINKED ISSUE #787: FDC3 Workbench: Create a separate version of Workbench for FDC3 2.0
## Enhancement Request

The workbench is intended to test the features of a specific FDC3 version and may not adapt well to a different version.

1. Create a separate version of the workbench for 2.0 development by introducing a new folder structure
    - Mimic the structure used by [fdc3-explained](https://github.com/ChartIQ/FDC3/tree/master/toolbox/fdc3-explained)
    - Move existing version into a 1.2 folder
    - Create a new version in a 2.0 folder
    - Add an index.html page to select correct version

1. Add a version check to both versions 
    - Highlight provider version number in red with a warning icon if the version does not match the workbench's version
      image

    - On mouseover of the warning icon ⚠️ display message (perhaps just with alt tag): Your FDC3 version (1.2) doesn't match the version of the FDC3 Workbench you are using (2.0)

### LINKED ISSUE #788: FDC3 Workbench: Add Support for FDC3 2.0 features
## Enhancement Request

The FDC3 Workbench should be updated to support 2.0 features (after #787 is resolved)

1. **Use the new names for ` joinChannel()` ->`joinUserChannel()` and `getSystemChannels()`-> `getUserChannels()`**
    1. Update types used to 2.0 versions
    2. Change all function calls to use new names
    3. Change tab names and other labels to replace <strong>System Channels</strong> with <strong>User Channels</strong>
    
2. **App's own metadata**
    1. Add an additional row to the FDC3 implementation metadata top-right, to display the appID
    2. Apply CSS to the value field to limit its width: 
        ```
        {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            overflow-wrap: anywhere;
            max-width: 150px;
        }
        ```
        
3. **Originating app metadata**
    1. In 2.0 Context and Intent Handler functions may be passed an additional metadata argument identifying the app that originated the message.
        1. See: [https://fdc3.finos.org/docs/next/api/ref/Types#contexthandler](https://fdc3.finos.org/docs/next/api/ref/Types#contexthandler) 
    1. Add a second element under each 'LAST RECEIVED CONTEXT' text area (User Channels, App Channels and Intents sections), with title 'RECEIVED FROM'
    1. If present the metadata argument is present, format it manually (to avoid {} taking up a line each) and populate the box with it, e.g.:  
        ```
        appId: someApp
        instanceId: f5300a97-c0df-4df2-922b-6cdc09db7a9e
        ```
    1. If NOT present, populate box with string 'METADATA NOT PROVIDED''

4. **Receiving an Intent result**
    1. In the Intents tab in left hand panel, in the two 'raise' elements ('Raise Intent' and Raise Intent for context'), after the intent is raised add a new 'resolution' element that will render the 'IntentResolution' we receive back from the raised intent (see [https://fdc3.finos.org/docs/next/api/ref/Metadata#intentresolution](https://fdc3.finos.org/docs/next/api/ref/Metadata#intentresolution) for details of the object returned)
    1. Render data:
        1. Resolved by (<code>resolution.source</code>):
            ```
            appId: someApp
            instanceId: f5300a97-c0df-4df2-922b-6cdc09db7a9e
            ```
        3. Result: pending…
    1. Await any result with: 
        ```
        //Raise an intent and retrieve a result from the IntentResolution</code></strong>
        let resolution = await agent.raiseIntent("intentName", context);
        try {
          const result = await resolution.getResult();
          //detect whether the result is Context or a Channel
          if (result && result.broadcast) { 
            //render channel
          } else if (result){
            //render context
          } else {
            //render no result
          }
        } catch(error) {
          console.error(`${resolution.source} returned a result error: ${error}`);
        }
        ```
    1. Render results appropriately:
        1. **Context:**
            Format with `JSON.stringify(context, null, 2)` and render in disabled text field (as used for 'Last received context' in right hand panel) and title 'RESULT CONTEXT'
        1. **Channel:**
        Check type of channel (`result.type == "app"` or `result.type == "private"`, see [https://fdc3.finos.org/docs/next/api/ref/Channel](https://fdc3.finos.org/docs/next/api/ref/Channel) for details)
            1. If type of channel is "app" 
            Add elements that are normally displayed on app channels tab after a channel is retrieved with <strong><code>fdc3.getOrCreateChannel</code></strong>
            1. If type of channel is "private", add the same elements
                1. Additionally, add a 'disconnect' button, that should call the disconnect function of a PrivateChannel: see [https://fdc3.finos.org/docs/next/api/ref/PrivateChannel](https://fdc3.finos.org/docs/next/api/ref/PrivateChannel) for details
                1. Add a 'Channel id' element to render the private channel's ID.
                1. In the right-hand panel, listeners tab, add a 'Private Channels' section where an element should be added when a listener is added to the channel and contexts received should be rendered (as they are for App Channels).
                    1. Channel will likely have an unfriendly <code>id</code>, which needs to be displayed so it can be (manually) connected back to the result received
                1. Automatically supply handlers for the 3 additional listener functions that private channels have:
                    2. `onAddContextListener(handler: (contextType?: string) => void): Listener;`
                    3. `onUnsubscribe(handler: (contextType?: string) => void): Listener;`
                    4. `onDisconnect(handler: () => void): Listener;`
                1. Render events received on these 3 handlers in the left-hand panel under the result element NOT in the right-hand panel where contexts received are rendered.
                    1. Treat these events like a log and accumulate them.
        1. **No result:**
        Simple string: No result returned
    1. Add a 'Clear result' button, that should delete the specific result element

5. **Sending an Intent Result**

    **Note:** The workbench will not be able to support complex intent result behaviors, but should be able to support example intent result behaviors to fulfill its purpose of enabling development and test. 

    1. On the intents tab, under 'Add Intent Listener', new options will be needed - they should be enabled via a checkbox: 
    ☑️  Send Intent Result 
    1. There are two types of result, which should be selected via radio buttons, each will have options: 
    🔘 Context result   | Context template dropdown 🔽 |   
    🔘 Channel result
      🔲  Use a private channel
      |--Channel name–-------|
      | my channel name&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
      |--------------------------|  
      ☑️  Send context result over channel
          | Context template dropdown 🔽 |

    **Note:**  The options selected should affect the code on the copy button next to the 'add listener' button \

    1. The context result option should be fairly simple, the intent handler should just return the selected context template.
    1. The Channel result is slightly more complex…
        1. If 'use a private channel' is selected, the intent handler registered should: 
            1. Hide or disable the channel name box
            1. Create a private channel with `fdc3.createPrivateChannel()`
            1. If 'Send context result over channel' is selected
                1. Add a listener to it with `channel.onAddContextListener(...)`
                1. Broadcast the selected context via the channel object when the listener fires.
            1. Return the (private) channel object
        1. If 'use a private channel' is NOT selected, the intent handler registered should: 
            1. Show/enable the 'channel name box'
            1. Retrieve the channel object with `fdc3.getOrCreateChannel(channelName)`
            1. Set-up a setTimeout call to wait 1.5 seconds then broadcast the select context bia the channel object
            1 Return the (app) channel object.

    **Note:**  We may later decide that the delay before sending context should be configurable (but isn't needed for private channels that sync on the addition of the context listener). We may also decide that we need to be able send multiple contexts with a delay between, which will require further development to set up the sequence with delays…

6. **Targeting a raised intent (at an app instance)**

    **Note:**  Intents can be targeted at particular appIds in FDC3 1.2 and appId/instanceId pairs in 2.0. However, there is no way to retrieve the list of available appIds in FDC3, hence we will have to rely on user data entry for the AppId. Instance Ids can be retrieved for an appId however.

    1. On the intents tab, add a check box under 'Raise Intent' and 'Raise Intent for Context' (above where results will be displayed, but below other controls): 
      ☑️  Target an app
             |--AppId–-----------|
             | myAppId&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
             |---------------------|
             ☑️  Target an instance of the app 
                   | Instance id drop down 🔽 | 

    1. Due to the space this will occupy, it is recommended that the fields are displayed when the checkboxes are clicked.
    1. Instance id drop down can be populated via a call to`fdc3.findInstances(...)`
    See: [https://fdc3.finos.org/docs/next/api/ref/DesktopAgent#findinstances](https://fdc3.finos.org/docs/next/api/ref/DesktopAgent#findinstances) 
    1. When the Raise Intent button is clicked, an object with the `appId`(and `instanceId`, if selected) should be passed as an argument
    See: [https://fdc3.finos.org/docs/next/api/ref/DesktopAgent#raiseintent](https://fdc3.finos.org/docs/next/api/ref/DesktopAgent#raiseintent)

### PR DESCRIPTION
**Pull request by kriswest**
resolves #787
resolves #788

An updated version of the FDC3 workbench for FDC3 2.0, which now includes:
- the ability to switch between 1.2 and 2.0 versions of the workbench
- the ability to target `raiseIntent` and `raiseIntentForContext`, using options gathered via `findIntent` and `findIntentByContext`
- the ability to set up intentResults in intent listeners - including both context and channel results, channel results can have contexts broadcast on to the channel
- a number of bug fixes, including polyfill of the node process module used in the schema validator
- tightened up layout in header and contexts tab
- updated schema URLs pointing to 'next'
- updated some website dependencies to improve the npm-audit results
- updated the redocly CDN URL for the appD page

Still in need of an update to MUI v5 (one day)

### CODE REVIEW

**Comment by kriswest** [comment]
@TheJuanAndOnly99 @robmoffat I believe the EasyCLA error on this PR:

image

is down to an EasyCLA bug. The two commits reported as having a missing user are linked to two users - they are merge commits for PRs into the PR branch, raised by one user and then altered and merged by another:

image

image

These are the only two commits in the PR linked to multiple users.

**Review by greyseer256 — COMMENTED** [review]
Code looks good!

Two things I noticed while testing, the 'All' context listener type didn't seem to register anything when I was broadcasting instruments, I know it used to, is that expected?

Also the right hand panel under the channels tab still says "System Channels" should that say "User Channels" now?

image

File: toolbox/fdc3-workbench/src/fixtures/logMessages.ts [hunk 1]
```diff
@@ -12,36 +12,36 @@ export const getLogMessage = (name: logMessagesName, type: logMessagesType, valu
 			error: `The FDC3 API is not ready${value ? ` (${value})` : ""}`,
 		},
 		getChannels: {
-			success: `Successfully retrieved System channels`,
+			success: `Retrieved System channels`,
```
[the remark this answers is not in the document]
Comment (kriswest) [reply toolbox/fdc3-workbench/src/fixtures/logMessages.ts:15]: SHould be updated - will fix! thx.

File: toolbox/fdc3-workbench/src/components/Header.tsx [hunk 2]
```diff
@@ -39,38 +46,97 @@ const useStyles = makeStyles((theme: Theme) =>
...
+					}
+					if (implInfo.fdc3Version) {
+						displayInfo.fdc3Version = implInfo.fdc3Version;
+					}
+					if (implInfo.appMetadata) {
+						displayInfo.appMetadata = {
+							appId: implInfo.appMetadata.appId ? implInfo.appMetadata.appId : "not specified",
+							version: implInfo.appMetadata.version ? implInfo.appMetadata.version : "not specified",
+						};
+					}
```
[the remark this answers is not in the document]
Comment (kriswest) [reply toolbox/fdc3-workbench/src/components/Header.tsx:118]: Yes we could!

**Review by mattjamieson — APPROVED** [review]
LGTM!

**Pull request merged**